### PR TITLE
Change configure Jupyter server steps from async to sync

### PR DESCRIPTION
**Pull request by lucyzhang929**
This PR changes some of the configure Jupyter server steps from async to sync to improve the Python kernel loading time. On Windows, the loading time decreased by 1.3 seconds on average. (No difference seen on Mac as the configuration step only took a couple of ms on Mac).

The configuration steps are creating and copying small folders/files. The abundance of async calls likely caused the configuration step to sometimes take up to 3 seconds on Windows while other times only taking <1 second.

### CODE REVIEW

File: extensions/notebook/src/common/utils.ts [hunk 1]
```diff
@@ -251,6 +259,15 @@ export async function exists(path: string): Promise<boolean> {
 	}
 }
 
+export function existsSync(path: string): boolean {
```
Comment (corivera) [inline extensions/notebook/src/common/utils.ts:262]: Why do we need this when we have fs.pathExistsSync() ?

**Review by Charles-Gagnon — COMMENTED** [review]
Is this a known issue with the async functions or something? I'm not sure why these would affect the startup time that much since they should be making the same OS calls in the end.

In general we should avoid using sync functions like this (especially for OS stuff) since it ties up the runtime thread while it's executing. 

What led you to the conclusion that this is the correct fix and what kind of testing did you do to verify this?

**Comment by lucyzhang929** [comment]
So I’m not sure if this is a known issue for async functions but this was one of @chlafreniere's theories for why the configuration step may be taking so long. 

In terms of testing, I got the average time improvement from a smoke test that opens up ADS and a python notebook 100 times and measures the time it takes for the Python kernel to be ready. In addition to that, I did some time measurements on both Mac and Windows and each individual sync call took <10ms, which I think shouldn’t have a negative impact on the rest of ADS right? The files being copied are in the extensions/notebook/ folder and are all 10 lines or less. But I guess if these files were to be moved or increased in size this could be problematic..

Definitely makes sense to avoid sync functions and we probably need a little more discussion/testing before making this change. @chlafreniere @Charles-Gagnon any suggestions on other tests I can do or maybe ideas on other problems that might be causing this wide range of time for the configuration step?

File: extensions/notebook/src/common/utils.ts (hunk 1 above)
Comment (lucyzhang929) [reply extensions/notebook/src/common/utils.ts:262]: Previously the exists method (line 252) was being called - not sure why fs.pathExists() wasn't used. But I can change it and remove this method.

**Comment by Charles-Gagnon** [comment]
Were you measuring the specific time taken for the file system calls? What might be happening is that something else is taking up the thread while the async functions are running which is increasing the time it takes to actually do the entire thing since it's having to switch contexts a lot more.

I'd be curious to have you run some perf tests on just doing single fs calls in a loop and measuring the time for the async vs sync to see if this is actually a perf difference in the call or if it's something else that is happening while the async functions are executing. 

Given that we're not strictly enforcing the async fs calls anyways (other extensions are already using them) and the importance to startup time here I don't really have a huge issue with doing this as long as we're sure of the improvement being made. It's unlikely to negatively affect the other extensions with what this is doing.

**Comment by lucyzhang929** [comment]
Yep I was measuring each mkdirSync and copySync call. I believe Chris had the same reasoning there (other calls causing the async functions here to be completed slower).
I can definitely run some perf tests - see if theres any perf difference between async vs sync.

**Comment by lucyzhang929** [comment]
Calling fs.copy and fs.copySync 5000 times each on a small text file resulted in copy taking about twice as long as copySync to complete. Seems like there is some overhead added to async calls in addition to the possibility of other things taking up the thread during the config jupyter step.

File: extensions/notebook/src/common/utils.ts [hunk 2]
```diff
@@ -30,6 +30,14 @@ export async function mkDir(dirPath: string, outputChannel?: vscode.OutputChanne
 		await fs.ensureDir(dirPath);
 	}
 }
+export function mkDirSync(dirPath: string, outputChannel?: vscode.OutputChannel): void {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/common/utils.ts:33]: Honestly - if you're looking for speed enhancements I'd get rid of the pathExistsSync check. ensureDirSync will only create the folder if it doesn't exist so checking that it doesn't first is kind of pointless. (you can decide if you care about the output channel message - I personally would just change it to something like "Ensuring folder {0} exists..."

The "check if file exists and if it doesn't create it" is generally considered an anti-pattern anyways. There's nothing preventing the folder from being created between the exists check and when you try to create it. So the suggested method is just to always try to create and then ignore any "already exists" errors.

Comment (corivera) [reply extensions/notebook/src/common/utils.ts:33]: Checking if a file exists in a separate call is even discouraged in the file system module documentation: https://nodejs.org/api/fs.html#fs_fs_stat_path_options_callback

"Using fs.stat() to check for the existence of a file before calling fs.open(), fs.readFile() or fs.writeFile() is not recommended. Instead, user code should open/read/write the file directly and handle the error raised if the file is not available." There are similar blurbs for other file checking methods.

Comment (lucyzhang929) [reply extensions/notebook/src/common/utils.ts:33]: Certainly makes a lot of sense - thanks for letting me know! I'll making the changes.

File: extensions/notebook/src/common/utils.ts [hunk 3]
```diff
@@ -23,20 +23,16 @@ export function getLivyUrl(serverName: string, port: string): string {
 }
 
 export async function mkDir(dirPath: string, outputChannel?: vscode.OutputChannel): Promise<void> {
-	if (!await fs.pathExists(dirPath)) {
-		if (outputChannel) {
-			outputChannel.appendLine(localize('mkdirOutputMsg', "... Creating {0}", dirPath));
-		}
-		await fs.ensureDir(dirPath);
+	if (outputChannel) {
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/common/utils.ts:26]: You can just shorten this to `outputChannel?.appendLine(....)`

File: extensions/notebook/src/common/utils.ts [hunk 4]
```diff
@@ -23,20 +23,16 @@ export function getLivyUrl(serverName: string, port: string): string {
...
+		outputChannel.appendLine(localize('ensureDirOutputMsg', "... Ensuring {0} exists", dirPath));
 	}
+	await fs.ensureDir(dirPath);
```
Comment (Charles-Gagnon) [inline extensions/notebook/src/common/utils.ts:29]: Would suggest renaming this function to `ensureDir` then to match the call being done.

**Review by Charles-Gagnon — APPROVED** [review]

**Review by corivera — APPROVED** [review]

**Review by chlafreniere — APPROVED** [review]
